src/silo/query_engine/actions/fasta.cpp

```diff
@@ -162,87 +167,148 @@ void Fasta::addSequencesToResultsForPartition(
+      query(connection, create_table_query);
+   }
 
+   // 4. Fill the decompressed sequences into the `QueryResultEntry` entries in `result`
    addSequencesFromResultTableToJson(
-      results, connection, result_table_name, sequence_names, database_partition, number_of_values
+      results, connection, result_table_name, sequence_names, database_partition, num_result_rows
    );
 
-   (void)query(connection, fmt::format("DROP TABLE {};", result_table_name));
-   (void)query(connection, fmt::format("DROP TABLE {};", key_table_name));
+   // XX don't we need guaranteed cleanup for those?
+   query(connection, fmt::format("DROP TABLE {};", result_table_name));
+   query(connection, fmt::format("DROP TABLE {};", key_table_name));
+
+   return last_row_id;
 }
 
+}  // namespace
+
+// Note: fasta_aligned.cpp has its own PARTITION_CHUNK_SIZE
+static const size_t PARTITION_CHUNK_SIZE = 10000;
+
 QueryResult Fasta::execute(const Database& database, std::vector<OperatorResult> bitmap_filter)
    const {
    for (const std::string& sequence_name : sequence_names) {
       CHECK_SILO_QUERY(
          database.unaligned_nuc_sequences.contains(sequence_name),
          "Database does not contain an unaligned sequence with name: '" + sequence_name + "'"
-      )
-   }
-
-   const std::string& primary_key_column = database.database_config.schema.primary_key;
-
-   size_t total_count = 0;
-   for (auto& filter : bitmap_filter) {
-      total_count += filter->cardinality();
+      );
    }
-   CHECK_SILO_QUERY(
-      total_count <= SEQUENCE_LIMIT,
-      fmt::format("Fasta action currently limited to {} sequences", SEQUENCE_LIMIT)
-   );
 
-   QueryResult results;
-   results.query_result.reserve(total_count);
-
-   for (uint32_t partition_index = 0; partition_index < database.partitions.size();
-        ++partition_index) {
-      const auto& database_partition = database.partitions[partition_index];
-      const auto& bitmap = bitmap_filter[partition_index];
-
-      addSequencesToResultsForPartition(results, database_partition, bitmap, primary_key_column);
-   }
+   size_t current_partition = 0;
+   // The unprocessed part of the result row numbers, counting from 0
+   // from the start of the current partition:
+   std::optional<Range<uint32_t>> remaining_result_row_indices;
+   return QueryResult{[sequence_names = sequence_names,
+                       remaining_result_row_indices,
+                       bitmap_filter =
+                          std::make_shared<std::vector<OperatorResult>>(std::move(bitmap_filter)),
+                       &database,
+                       current_partition](std::vector<QueryResultEntry>& results) mutable {
+      while (current_partition < database.partitions.size()) {
+         // We drain bitmap_filter as we process the query (because
+         // the only way to get an iterator over a bitmap is from the
+         // beginning?)! And `remaining_result_row_indices` is really
+         // only for show now.
```

Comment:
For the debug messages, via `result_row_indices`. But actually also for a bit more than that (also via `result_row_indices`). I'm not 100% satisfied because I think there's a slight duplication of state, that's why I added the comment, maybe I should make that a bit clearer.

Reply:
(A high quality iterator based infrastructure as mentioned above might avoid the issue of duplication of state; but seems out of reach.)